### PR TITLE
Generalized parser and table for cryptographic materials (eg: certificates)

### PR DESCRIPTION
**Pull request by directionless**
With `pkg/keyidentifier`, we've long had the ability to parse private keys. But we've often wanted to add certificates to that. Unfortunately that library and associated table hardcode several assumptions about the shape of the resultant data.

Thus, `pkg/cryptoinfo` and the associated `kolide_cryptoinfo`. This is designed to return nested data, suitable for being handed to dataflatten. 

Right now, it only parses certificates, though it should handle both common formats (DER and PEM). I think it has the bones to grow into handling all other keys as needed. We could replace `pkg/keyidentifier` with it. 

For example:
```
osquery> select count(*) from kolide_cryptoinfo where path = '/etc/ssl/cert.pem' ;
+----------+
| count(*) |
+----------+
| 7903     |
+----------+
osquery> select value from kolide_cryptoinfo where path = '/etc/ssl/cert.pem' and fullkey like '%/Subject/CommonName' limit 5;
+-----------------------------------------------------------+
| value                                                     |
+-----------------------------------------------------------+
| Autoridad de Certificacion Firmaprofesional CIF A62634068 |
| Chambers of Commerce Root - 2008                          |
| Global Chambersign Root - 2008                            |
| ACCVRAIZ1                                                 |
| Actalis Authentication Root CA                            |
+-----------------------------------------------------------+
```

### CODE REVIEW

File: pkg/cryptoinfo/parse_certificate.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		Subject:               c.Subject,
+		SubjectParsed:         c.Subject.String(),
+		URIs:                  c.URIs,
+		Version:               c.Version,
+	}, nil
+
+}
+
+var keyUsageBits = map[x509.KeyUsage]string{
+	x509.KeyUsageContentCommitment: "Content Commitment",
```
Comment (blaedj) [inline pkg/cryptoinfo/parse_certificate.go:77]: Is there a reason that you didn't include `x509.KeyUsageDigitalSignature:  "Digital Signature",` here?

File: pkg/cryptoinfo/info.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	return ki
+}
+
+func (ki *KeyInfo) SetData(data interface{}, err error) *KeyInfo {
+	ki.Data = data
+	ki.Error = err
+	return ki
+}
+
+func (ki *KeyInfo) MarshalJSON() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (directionless) [reply pkg/cryptoinfo/info.go:41]: This is used. It's a little hard to summarize why/how. 

So, this is meant to take different sorts of data and stitch it together. But go is strongly typed. So how to handle method signatures and data passing always requires a bit of cleverness.

One pattern is to pass around maps. I tried that at first, but it felt like it was getting hard to keep the keys unified across the namespaces. 

So instead I looked at creating this custom `KeyInfo` type. Everything in here creates KeyInfo structs, and the end result is `[]KeyInfo`. But, and this is where it's a bit magical, when we marshal this KeyInfo struct into useful end data, we _don't_ want it as written. I don't want `Error` if it's nil, and I want the `DataName` to be the key, with a the value of `Data`. 

I might be wrong? But the intent is to end up with data like:
```
{
  type: certificate,
  encoding: pem,
  certificate: ...
},
{
  type: key,
  encoding: pem,
  key: ...,
}
```

So to get there, `KeyInfo` has a custom json marshal function with some conditionals.

TL;DR: this controls how the marshal-to-json output looks

File: pkg/cryptoinfo/parse_certificate.go (hunk 1 above)
Comment (directionless) [reply pkg/cryptoinfo/parse_certificate.go:77]: Oversite. Or maybe it didn't exist when I first wrote this code. 
```suggestion
        x509.KeyUsageDigitalSignature: "Digital Signature",
	x509.KeyUsageContentCommitment: "Content Commitment",
```

File: pkg/cryptoinfo/info.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (blaedj) [reply pkg/cryptoinfo/info.go:41]: Gotcha, makes sense, thank you for explaining. I wasn't aware of (or maybe forgot) the trick of defining `UnmarshallJSON` to get the custom unmarshalling behavior. 

(Too much magic :trollface: )

[the remark this answers is not in the document]
Comment (directionless) [reply pkg/cryptoinfo/info.go:41]: I think the weirder part is that I'm doing this to encapsulate slightly different underlying data. Not sure if there's a more correct pattern. Maaaaybe go interfaces? Let the parse functions return whatever, as long as it confirms to an interface. I'm not really sure I want to try another rewrite

[the remark this answers is not in the document]
Comment (blaedj) [reply pkg/cryptoinfo/info.go:45]: Yea, I'm not sure tbh. I think that I probably won't have opinions until I try to change this code, or need to interact with it more.

[the remark this answers is not in the document]
Comment (directionless) [reply pkg/cryptoinfo/info.go:45]: Same!

**Review by blaedj — APPROVED** [review]
